src/main/java/com/google/graphgeckos/dashboard/storage/Builder.java

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.graphgeckos.dashboard.storage;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.Entity;
+
+/**
+ * Contains the information retrieved from a single build bot. It is used as a subclass
+ * of {@link #BuildInfo BuildInfo}, and for converting ParsedBuildbotData objects
+ * to Builder objects, used by BuildInfo.
+ */
+@Entity(name = "builder")
+public class Builder {
+
+  private final String builderName;
+
+  // The logs of each compilation stage, stored as described by {@link #Log Log}.
+  private final List<Log> logs;
+
+  // Builder compilation status, as described by {@link #BuilderStatus BuilderStatus}.
+  private final BuilderStatus status;
+
+  /**
+   * Converts a {@link #ParsedBuildbotData ParsedBuildbotData} object to a Builder object.
+   * The fields should be sanitized before this operation, as no field in Builder may be
+   * null.
```

Comment:
Consider checking these conditions in this constructor; throw IllegalArgumentException if there is a non-expected field that is null and add a @throws javadoc comment about that.

Reply:
This is already checked in DatastoreRepository. Should I remove explicitly checking for `validUpdateData` there, and instead check it here, and leave this exception as a fallthrough in DatastoreRepository?

Reply:
>Should I remove explicitly checking for validUpdateData there, and instead check it here, and leave this exception as a fallthrough in DatastoreRepository?

Yes, I think that would be good! In principle you could create an instance of this Builder class from outside of the DatastoreRepository implementation, and then it would be appropriate to explicitly check for `validUpdateData` here, in order to ensure the guarantees that this class provides (its getters should return non-nulls). 

But I'd suggest taking this a step further (in a follow-up PR): since the validity of a `ParsedBuildbotData` instance is an internal property of that class (the thing checked by `validUpdateData()`), I'd suggest to update ParsedBuildbotData's constructor to check its arguments and throw an IllegalArgumentException if the arguments would lead to producing an invalid instance. This way you ensure that all ParsedBuildbotData instances are valid, and you could eliminate the checks in client code such as here that consumes a ParsedBuildbotData instance and it expects it to be valid. Note that this will make it not possible to create invalid ParsedBuildbotData instances at runtime, but I think we don't strictly need to be able to represent such invalid instances.
Additionally, this would be better as it would identify cases where the update data is invalid earlier; this is known as fail-fast and is generally a good rule of thumb to try to fail and bring up a problem as early as it is encountered to help out with investigations.

Reply:
Thank you for the follow-up. The same fail-fast approach would go for ParsedGitData as well, so as to avoid any further checks inside the DatastoreRepository. Both of these will be part of an upcoming PR.

Reply:
#79 solves this issue. Should this JavaDoc be updated in any way following this change?

Reply:
>#79 solves this issue. Should this JavaDoc be updated in any way following this change?
Great! After #79 gets in, I'd update this sentence `The fields should be sanitized ...` to something like `"botData" must not be null.`, or just add a `@throws IllegalArgumentException if botData is null` section.